precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,615 @@
+	}
+
+	_, err = m.authzKeeper.DispatchActions(context.SdkCtx(), bridgeAddr, []sdk.Msg{msg})
+	if err != nil {
+		return nil, err
+	}
+
+	assetUnlocked, err := m.bridgeKeeper.AssetsUnlocked(
+		context.SdkCtx(),
+		inputs.Token.Bytes(),
+		sdkAmount,
+		uint8(inputs.Chain),
+		inputs.Recipient,
+	)
+	if err != nil {
+		// TODO(JEREMY): error happened here, funds might have been
+		// lost, should we panic?
+		return nil, fmt.Errorf("failed to send AssetsUnlocked to bridge: %w", err)
+	}
+
+	// finally update the journal entries to propagate the changes
+	// done to the gas token (BTC in our case)
+	balanceDelta, overflow := uint256.FromBig(inputs.Amount)
+	if overflow {
+		return nil, fmt.Errorf("conversion from big.Int to uint256.Int overflowed: %v", inputs.Amount)
+	}
+
+	// only one side of the transfer to update here as we
+	// burnt funds
+	journal := context.Journal()
+	journal.SubBalance(senderAddr, balanceDelta, tracing.BalanceChangeTransfer)
+
+	return assetUnlocked, nil
+}
+
+func (m *BridgeOutMethod) validateAuthorizationLimits(
+	sendAuth *banktypes.SendAuthorization,
+	requestedCoins sdk.Coins,
+) error {
+	if sendAuth.SpendLimit == nil || sendAuth.SpendLimit.Empty() {
+		return fmt.Errorf("no allowance for for %v", requestedCoins[0].Denom)
+	}
+
+	for _, requestedCoin := range requestedCoins {
+		allowedAmount := sendAuth.SpendLimit.AmountOf(requestedCoin.Denom)
+		if allowedAmount.IsZero() {
+			return fmt.Errorf("no allowance for %s", requestedCoin.Denom)
+		}
+		if requestedCoin.Amount.GT(allowedAmount) {
+			return fmt.Errorf(
+				"requested amount %s exceeds allowed amount %s for %s",
+				requestedCoin.Amount,
+				allowedAmount,
+				requestedCoin.Denom,
+			)
+		}
+	}
+
+	// Check allowed list if it exists
+	// It shouldn't be set seeing that we don't really use cosmos-sdk
+	// but just in case?
+	if len(sendAuth.AllowList) > 0 {
+		found := false
+		senderAddrStr := sdk.AccAddress(common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes()).String()
+
+		for _, allowedAddr := range sendAuth.AllowList {
+			if allowedAddr == senderAddrStr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("recipient address not in authorization allow list")
+		}
+	}
+
+	return nil
+}
+
+// validate applies more extensive validation to the inputs, not only
+// the semantic validation but also verify balances, and tokens etc.
+func (m *BridgeOutMethod) validate(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) error {
+	sdkCtx := context.SdkCtx()
+	// first check that the token is either BTC or a supported ERC20
+	if !m.isValidToken(sdkCtx, inputs.Token) {
+		return fmt.Errorf("unsupported token: %v", inputs.Token)
+	}
+
+	spenderAddr := context.MsgSender()
+	if err := m.isAmountSpendableByBridge(sdkCtx, spenderAddr, inputs.Token, inputs.Amount); err != nil {
+		return err
+	}
+
+	return m.isAmountAvailable(sdkCtx, spenderAddr, inputs.Token, inputs.Amount)
```

Comment:
Do we actually need to check allowance and balance before burning? For ERC20, the `burnFrom` call does those both checks underneath. Same for BTC as we are relying on the `x/auth` module for allowance and `x/bank` for burning and balance checking. I know the requirement of validating this was mentioned in the technical spec but I think implicit validation done by `burnFrom`/`x/auth`/`x/bank` totally fullfills it. WDYT? This will greatly simplify our code.

Reply:
Happy to remove the explicit check, I like having those explicitly covered but am also happy to completely remove them if you want.

Reply:
What are the benefits justifying the explicit checks? I see this may allow us to fail fast but other than that? Do you see anything else?

Regarding the downsides, the major one is more code and increased complexity so more places where we can do mistakes.

Sure, we need to trust that `burnFrom` and `x/bank` will actually check the allowance and balance but this is an inherent part of their logic and not doing so would mean the given ERC20 token or the `x/bank` error have a critical flaw and even if we check explicitly for allowance and balance, that wouldn't help much. Also, if we would be writing this code in pure Solidity, we probably wouldn't do those checks explicitly but just rely on `burnFrom`.

Reply:
Sounds good, just removing all these.

Reply:
I see this was addressed. Closing.